File: javaexample/src/main/java/io/chthonic/mythos/javaexample/ui/fragments/RoFragment.java
package io.chthonic.mythos.javaexample.ui.fragments;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import io.chthonic.mythos.javaexample.ui.presenters.RoPresenter;
import io.chthonic.mythos.javaexample.ui.vus.RoVu;
import io.chthonic.mythos.mvp.MVPDispatcher;
import io.chthonic.mythos.mvp.MVPFragment;
import io.chthonic.mythos.viewmodel.PesenterCacheViewModel;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function4;

/**
 * Created by jhavatar on 3/22/2016.
 */
public class RoFragment extends MVPFragment<RoPresenter, RoVu> {

    public static final String TAG = RoFragment.class.getSimpleName();
    private static final int MVP_UID = TAG.hashCode();

    @NonNull
    @Override
    public MVPDispatcher<RoPresenter, RoVu> createMVPDispatcher() {
        return new MVPDispatcher<>(MVP_UID,
                PesenterCacheViewModel.getViewModelPresenterCache(this, MVP_UID, new Function0<RoPresenter>() {
                    @Override
                    public RoPresenter invoke() {
                        return new RoPresenter();
                    }
                }),
                new Function4<LayoutInflater, Activity, Fragment, ViewGroup, RoVu>() {
                    @Override
                    public RoVu invoke(LayoutInflater layoutInflater, Activity activity, Fragment fragment, ViewGroup parentView) {
                        return new RoVu(layoutInflater, activity, fragment, parentView);
                    }
                });
    }
}

File: kotlinexample/src/main/kotlin/io/chthonic/mythos/kotlinexample/ui/fragments/RoFragment.kt
package io.chthonic.mythos.kotlinexample.ui.fragments

import android.os.Bundle
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import io.chthonic.mythos.kotlinexample.App
import io.chthonic.mythos.kotlinexample.ui.presenters.RoPresenter
import io.chthonic.mythos.kotlinexample.ui.vus.RoVu
import io.chthonic.mythos.mvp.MVPDispatcher
import io.chthonic.mythos.mvp.MVPFragment
import io.chthonic.mythos.mvp.PresenterCacheLoaderCallback

/**
 * Created by jhavatar on 3/22/2016.
 */
class RoFragment : MVPFragment<RoPresenter, RoVu>() {

    companion object {
        val TAG: String by lazy {
            RoFragment::class.java.simpleName
        }

        private val MVP_UID by lazy {
            TAG.hashCode()
        }

        val lIFECYCLE_KEY = "ro"
    }

    private lateinit var  refwatcher: RefWatcher

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        refwatcher = LeakCanary.install(App.app)
    }

    override fun createMVPDispatcher(): MVPDispatcher<RoPresenter, RoVu> {
        return MVPDispatcher(MVP_UID,
                PresenterCacheLoaderCallback(this.activity!!, { RoPresenter() }),
                ::RoVu)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        refwatcher.watch(this)
    }

}